crates/cargo-winrt/src/main.rs

```diff
@@ -284,14 +284,14 @@ impl DependencyDescriptor {
         match self {
             DependencyDescriptor::NugetOrg { name, version } => {
                 let url = format!("https://www.nuget.org/api/v2/package/{}/{}", name, version);
-                let bytes = try_download(url, 5)?;
+                let bytes = try_download(url, 5, true)?;
```

Comment:
I'm confused why nuget packages include the body by default and downloading from url does not. I would expect the opposite to be true... Can you explain?

Reply:
If you make a debug here to print header:

```
diff --git a/crates/cargo-winrt/src/main.rs b/crates/cargo-winrt/src/main.rs
index 40b2fd0..103b592 100644
--- a/crates/cargo-winrt/src/main.rs
+++ b/crates/cargo-winrt/src/main.rs
@@ -349,6 +349,7 @@ fn try_download(url: String, recursion_amount: u8, include_body: bool) -> anyhow
         transfer
             .header_function(|header| {
                 let header = std::str::from_utf8(header).unwrap();
+                println!("header: {}", header);
                 if header.starts_with("HTTP/1.1 ") {
                     let n = &header[9..12];
                     *status = Some(n.parse::<u16>().expect("Should be number"));
```

You will see:

```
    Fetching Microsoft.AI.MachineLearning
	
https://www.nuget.org/api/v2/package/Microsoft.AI.MachineLearning/1.3.0
	
header: HTTP/1.1 302 Found
header: Cache-Control: private
header: Transfer-Encoding: chunked
header: Content-Type: text/html; charset=utf-8
header: Location: https://globalcdn.nuget.org/packages/microsoft.ai.machinelearning.1.3.0.nupkg
header: Request-Context: appId=cid-v1:338f6804-b1a9-4fe3-bba7-c93064e7ae7b
header: Access-Control-Expose-Headers: Request-Context
header: Content-Security-Policy: frame-ancestors 'none'
header: X-Frame-Options: deny
header: X-XSS-Protection: 1; mode=block
header: X-Content-Type-Options: nosniff
header: Strict-Transport-Security: max-age=31536000
header: Set-Cookie: ARRAffinity=e56c4413884e44fa2949035523b026d52af6ee3f783d256e3cbeb42fa95d6022;Path=/;HttpOnly;Domain=nuget-prod-v2gallery-appservice.trafficmanager.net
header: Date: Tue, 21 Jul 2020 01:18:30 GMT
header:

https://globalcdn.nuget.org/packages/microsoft.ai.machinelearning.1.3.0.nupkg

header: HTTP/1.1 200 OK
header: Accept-Ranges: bytes
header: Access-Control-Allow-Origin: *
header: Access-Control-Expose-Headers: x-ms-request-id,Server,x-ms-version,Content-Length,Date,Transfer-Encoding
header: Cache-Control: max-age=86400
header: Content-MD5: AAB1USeDetROmk8fD/XLqg==
header: Content-Type: application/octet-stream
header: Date: Tue, 21 Jul 2020 01:18:32 GMT
header: Etag: 0x8D7FB804A83621E
header: Expires: Wed, 22 Jul 2020 01:18:32 GMT
header: Last-Modified: Mon, 18 May 2020 23:07:52 GMT
header: Server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0
header: Strict-Transport-Security: max-age=31536000
header: X-Cache: HIT
header: X-CDN-Rewrite: Root path in dist
header: X-Content-Type-Options: nosniff
header: x-ms-blob-type: BlockBlob
header: x-ms-lease-status: unlocked
header: x-ms-meta-da7b2905_0f3c_4262_921c_b1593d1336f1_ESRP_RequestId: 8de9e9d0-befe-491f-849c-7e0fd05eba05
header: x-ms-meta-SHA512: gxrx74NSZh8TcDRLwYO9MqdTfn/5NfMVOMlAmdSMzyVNNFKogQWQVxsZxLlp2oIuDu0uByUQ0PNPjmFvk3ndLQ==
header: x-ms-request-id: 30fe239d-101e-0015-38fc-5ef2cc000000
header: x-ms-version: 2009-09-19
header: Content-Length: 172281918
header:
```

First time try_download from https://www.nuget.org/api/v2/package/Microsoft.AI.MachineLearning/1.3.0, it would include body. As it will cause a 302 redirect, otherwise it will cause 404

Second time it try to reach real url https://globalcdn.nuget.org/packages/microsoft.ai.machinelearning.1.3.0.nupkg, for now it should not include body, otherwise it will stick there to download both header and body.

Reply:
I'm sorry but I still don't get what the problem is. The first request results in a 302 response telling the caller to use a different URL. Even if the body is downloaded, it should at most contain a few bytes of html describing the redirect but not the entire nuget package.

Reply:
> I'm sorry but I still don't get what the problem is. The first request results in a 302 response telling the caller to use a different URL. Even if the body is downloaded, it should at most contain a few bytes of html describing the redirect but not the entire nuget package.

Hi Kenny, that's right, to get 302 response is include body that's fine, and we must include body to let 302 happen. My concern is 200 response after 302. To get 200 from the redirected URL and if it include body it will cause much time to download body when just want header.

I push a test repo here: https://github.com/uhuntu/curl-rust.git

You can just run a sample I wrote by: ```cargo run --example test_hunt```

You will measure that if I don't set ```handle.nobody(true).unwrap();```, the program should be seeing never end in low bandwidth environment, but actually I just want header. If you put ```handle.nobody(true).unwrap();``` back, the program will quit immediately after it get header.

```
use std::str;

fn main() {
    let mut handle = curl::easy::Easy::new();
    handle.url("https://globalcdn.nuget.org/packages/microsoft.ai.machinelearning.1.3.0.nupkg").unwrap();
    // handle.nobody(true).unwrap();
    handle.header_function(|header| {
        print!("header: {}", str::from_utf8(header).unwrap());
        true
    }).unwrap();
    handle.perform().unwrap();
}
```

Reply:
I think I'm following now. The code logic in [`try_download`](https://github.com/microsoft/winrt-rs/blob/6ab99ef9df3840caadc1f0ed9c82ea78b208a8e6/crates/cargo-winrt/src/main.rs#L441) first requests the resource (including the body), evaluates the HTTP response code, and then either follows the redirect (302) or re-initiates the transfer (200), including the body. Either way, the body from the first request gets dropped. The net result is that the body gets downloaded twice.

That is indeed wasteful, although I'm skeptical that the proposed change is the right way to address this issue: For one, it imposes the assumption, that the first 302 response is never followed by another 302. While I don't know whether HTTP allows for that, it seems plausible that it would. Worse, though, the proposed change doesn't help in situations, where the *first* request already returns a 200 code.

Moving the `write_function` from the [`200u16` match arm](https://github.com/microsoft/winrt-rs/blob/6ab99ef9df3840caadc1f0ed9c82ea78b208a8e6/crates/cargo-winrt/src/main.rs#L468) into the initial request (or otherwise preserving the downloaded document) would address both of these issues.

Reply:
I just re-submit the code as @tim-weis mentioned.